### PR TITLE
Add JetBrains RustRover IDE Supportive

### LINKED ISSUE #17480: [IDE] RustRover IDE Support
### The feature request

Recently, a new IDE from Jet Brains was launched called RustRover, and I hope that Git Desktop can support this compiler.

### Proposed solution

Just recognize [RustRover](https://www.jetbrains.com/rust/) like other ides

### Additional context

image

### PR DESCRIPTION
**Pull request by SnowballXueQiu**
Closes #17480

## Description
- Add JetBrains RustRover IDE Supportive

### Screenshots

image
image

### CODE REVIEW

**Review by SnowballXueQiu — COMMENTED** [review]
👂

**Comment by SnowballXueQiu** [comment]
@steveward 👀

File: docs/technical/editor-integration.md [hunk 1]
```diff
@@ -39,6 +39,7 @@ These editors are currently supported:
  - [JetBrains PyCharm](https://www.jetbrains.com/pycharm/)
  - [JetBrains RubyMine](https://www.jetbrains.com/rubymine/)
  - [JetBrains GoLand](https://www.jetbrains.com/go/)
+- [JetBrains RustRover](https://www.jetbrains.com/rust/)
```
Comment (zdimension) [inline docs/technical/editor-integration.md]: missing indent here

Comment (SnowballXueQiu) [reply docs/technical/editor-integration.md:42]: what ident？

Comment (SnowballXueQiu) [reply docs/technical/editor-integration.md]: how i need indent？ i think this is normal indent👉👈

Comment (SnowballXueQiu) [reply docs/technical/editor-integration.md]: ok i understood yours, thx!

**Review by SnowballXueQiu — COMMENTED** [review]
👍

**Comment by sergiou87** [comment]
Thanks for your contribution @SnowballXueQiu 

We haven't been able to make this work on Windows. Could you try it yourself and make sure it does? Thank you!

**Comment by SnowballXueQiu** [comment]
> Thanks for your contribution @SnowballXueQiu
> 
> We haven't been able to make this work on Windows. Could you try it yourself and make sure it does? Thank you!

oh, you can try to use the following command: 'yarn compile:prod && cross-env NODE_ENV=production ts-node -P script/tsconfig.json script/build.ts', then after build, i got the dist folder output.

and you can run the GitHubDesktop.exe on 'dist\GitHubDesktop-win32-x64\'

if you also have any questions, you can reply me and i'll tell you.

thanks!

**Comment by SnowballXueQiu** [comment]
as following

image

image

**Comment by SnowballXueQiu** [comment]
addition: through my tests, it can normally open the RustRover Softwave after i click 'Open in Jetbrains RustRover'

**Comment by SnowballXueQiu** [comment]
@sergiou87

**Comment by sergiou87** [comment]
@SnowballXueQiu it doesn't work here 😞 

image

Where did you get RustRover from? I got the latest preview version and the problem is the naming of the Registry key doesn't match the usual pattern for JetBrains IDEs:
2023-10-27 at 10 17

I also see your second screenshot is from macOS. Have you tested this on Windows?

**Comment by SnowballXueQiu** [comment]
> @SnowballXueQiu it doesn't work here 😞 
> 
> 
> 
> image
> 
> 
> 
> Where did you get RustRover from? I got the latest preview version and the problem is the naming of the Registry key doesn't match the usual pattern for JetBrains IDEs:
> 
> 2023-10-27 at 10 17
> 
> 
> 
> I also see your second screenshot is from macOS. Have you tested this on Windows?

im my computer the RustRover is succeed in identifying and launching. Idk why cant identify in yours at present.

And all of the screenshots are from the windows not macos.

**Comment by SnowballXueQiu** [comment]
also cant appear results？
@sergiou87

**Comment by SnowballXueQiu** [comment]
> @SnowballXueQiu it doesn't work here 😞
> 
> image
> 
> Where did you get RustRover from? I got the latest preview version and the problem is the naming of the Registry key doesn't match the usual pattern for JetBrains IDEs: 2023-10-27 at 10 17
> 
> I also see your second screenshot is from macOS. Have you tested this on Windows?

i ckecked out the Reg config, and found that the Registry name is right for it, just like following:

image

image

**Comment by Birditch** [comment]
@sergiou87 
When I try to clone this repository and compile it.
Use the `build:prod`script in `package.json` to compile.
The following compile success prompt appears:
```
Built to /Users/nothing/Documents/Projects/desktop/dist/GitHub Desktop-darwin-arm64
✨ Done in 33.57s.
```
Then I tried to go into this directory and opened `GitHub Desktop`
Then go to the project which is shown below.
image
You can easily see `RustRover` in the `External Editor` list.

Tips:
- My computer is a MacBook Pro M2 Max 2023.
- The cloned URL is `https://github.com/SnowballXueQiu/desktop.git`.
-  I installed my `RustRover` via `Jetbrains Tool Box`.
- The time is a few minutes before posting this reply.

In summary, you may have a problem with your own testing environment? Perhaps you should try retesting or checking that the `RustRover` environment is working.

**Comment by sergiou87** [comment]
Oh, so maybe installing it from the Toolbox is what makes the difference. I installed it from https://www.jetbrains.com/rust/ directly. In your case it's picking RustRover using the `jetBrainsToolboxScriptName` field of the editor config and not the other ones.

I will try to test it later with the Toolbox version. Thanks for helping out to clarify the issue!

**Comment by SnowballXueQiu** [comment]
> Oh, so maybe installing it from the Toolbox is what makes the difference. I installed it from https://www.jetbrains.com/rust/ directly. In your case it's picking RustRover using the `jetBrainsToolboxScriptName` field of the editor config and not the other ones.
> 
> 
> 
> I will try to test it later with the Toolbox version. Thanks for helping out to clarify the issue!

idk how to adapt to RustRover downloaded from Official Website. Can you tell me the specific name about it？ then i will make it change or you can add it by yourself thx.

**Comment by SnowballXueQiu** [comment]
👀

**Comment by SnowballXueQiu** [comment]
> 👀

so there're any ideas？ @sergiou87

File: app/src/lib/editors/win32.ts [hunk 2]
```diff
@@ -498,6 +498,14 @@ const editors: WindowsExternalEditor[] = [
     displayNamePrefixes: ['Pulsar'],
     publishers: ['Pulsar-Edit'],
   },
+  {
+    name: 'JetBrains RustRover',
+    registryKeys: registryKeysForJetBrainsIDE('RustRover'),
```
Comment (tsvetilian-ty) [inline app/src/lib/editors/win32.ts:503]: The `registryKeysForJetBrainsIDE` automatically generates possible registry keys for major and minor stable versions of JetBrains products. Currently, RustRover is in [EAP](https://www.jetbrains.com/resources/eap/), making it unstable from a registry key point of view. 

You can't currently use the `registryKeysForJetBrainsIDE` for RustRover!

File: app/src/lib/editors/darwin.ts [hunk 3]
```diff
@@ -220,6 +220,10 @@ const editors: IDarwinExternalEditor[] = [
     name: 'Cursor',
     bundleIdentifiers: ['com.todesktop.230313mzl4w4u92'],
   },
+  {
+    name: 'RustRover',
```
Comment (tsvetilian-ty) [inline app/src/lib/editors/darwin.ts:224]: It's a good idea to mention that RustRover is EAP and not stable!

File: docs/technical/editor-integration.md [hunk 4]
```diff
@@ -39,6 +39,7 @@ These editors are currently supported:
  - [JetBrains PyCharm](https://www.jetbrains.com/pycharm/)
  - [JetBrains RubyMine](https://www.jetbrains.com/rubymine/)
  - [JetBrains GoLand](https://www.jetbrains.com/go/)
+ - [JetBrains RustRover](https://www.jetbrains.com/rust/)
```
Comment (tsvetilian-ty) [inline docs/technical/editor-integration.md:42]: I see that you updated the docs for Windows, but there is a list with editors for each platform!

**Comment by tsvetilian-ty** [comment]
Hi @SnowballXueQiu 👋 

@sergiou87 is correct in [this comment](https://github.com/desktop/desktop/pull/17481#issuecomment-1784604962)!

You see the RustRover because you have it installed through Toolbox, and `jetBrainsToolboxScriptName` is detecting the version, but if you remove the version from Toolbox and install the executable directly from [the JetBrains website](https://www.jetbrains.com/rust/nextversion/) you won't see it ([more info in this comment I left](https://github.com/desktop/desktop/pull/17481#discussion_r1389693985))

**Comment by SnowballXueQiu** [comment]
> Hi @SnowballXueQiu 👋
> 
> @sergiou87 is correct in [this comment](https://github.com/desktop/desktop/pull/17481#issuecomment-1784604962)!
> 
> You see the RustRover because you have it installed through Toolbox, and `jetBrainsToolboxScriptName` is detecting the version, but if you remove the version from Toolbox and install the executable directly from [the JetBrains website](https://www.jetbrains.com/rust/nextversion/) you won't see it ([more info in this comment I left](https://github.com/desktop/desktop/pull/17481#discussion_r1389693985))

So you mean that, I can only wait for the official version at the moment? Or is there another way to solve it？

**Comment by tsvetilian-ty** [comment]
> So you mean that, I can only wait for the official version at the moment? Or is there another way to solve it？

I mean that you can't use the `registryKeysForJetBrainsIDE` because it won't catch the proper registry entry.

**A possible solution:**
_I haven't tried it, so it's just an idea!_

From [the screenshot](https://github.com/desktop/desktop/pull/17481#issuecomment-1782497862) that @sergiou87 shared, we can see that for the EAP version, they are using the [build number](https://www.jetbrains.com/rust/nextversion/) as part of the registry key, RustRover 233.8264.32

You could try to use `Wow64LocalMachineUninstallKey` and `CurrentUserUninstallKey` to get a specific version. ([check editor-integration.md](https://github.com/desktop/desktop/blob/development/docs/technical/editor-integration.md))

In conclusion, you could try to do something to cover  some versions :
```typescript
  {
    name: 'JetBrains RustRover EAP',
    registryKeys: [
      Wow64LocalMachineUninstallKey('RustRover 233.10527.39'), 
      Wow64LocalMachineUninstallKey('RustRover 233.8264.32'),
      Wow64LocalMachineUninstallKey('RustRover 233.8264.22'),
      
      CurrentUserUninstallKey('RustRover 233.10527.39'),
      CurrentUserUninstallKey('RustRover 233.8264.32'),
      CurrentUserUninstallKey('RustRover 233.8264.22'),
    ],
    executableShimPaths: executableShimPathsForJetBrainsIDE('rustrover'),
    jetBrainsToolboxScriptName: 'rustrover',
    displayNamePrefixes: ['RustRover '],
    publishers: ['JetBrains s.r.o.'],
  },
```
> [!WARNING]  
> The problem is that this solution is not scalable since the EAP versions are frequently updated.

@sergiou87, what are your thoughts on the subject?

**Comment by SnowballXueQiu** [comment]
> > So you mean that, I can only wait for the official version at the moment? Or is there another way to solve it？
> 
> 
> 
> I mean that you can't use the `registryKeysForJetBrainsIDE` because it won't catch the proper registry entry.
> 
> 
> 
> **A possible solution:**
> 
> _I haven't tried it, so it's just an idea!_
> 
> 
> 
> From [the screenshot](https://github.com/desktop/desktop/pull/17481#issuecomment-1782497862) that @sergiou87 shared, we can see that for the EAP version, they are using the [build number](https://www.jetbrains.com/rust/nextversion/) as part of the registry key, RustRover 233.8264.32
> 
> 
> 
> You could try to use `Wow64LocalMachineUninstallKey` and `CurrentUserUninstallKey` to get a specific version. ([check editor-integration.md](https://github.com/desktop/desktop/blob/development/docs/technical/editor-integration.md))
> 
> 
> 
> In conclusion, you could try to do something to cover  some versions :
> 
> ```typescript
> 
>   {
> 
>     name: 'JetBrains RustRover EAP',
> 
>     registryKeys: [
> 
>       Wow64LocalMachineUninstallKey('RustRover 233.10527.39'), 
> 
>       Wow64LocalMachineUninstallKey('RustRover 233.8264.32'),
> 
>       Wow64LocalMachineUninstallKey('RustRover 233.8264.22'),
> 
>       
> 
>       CurrentUserUninstallKey('RustRover 233.10527.39'),
> 
>       CurrentUserUninstallKey('RustRover 233.8264.32'),
> 
>       CurrentUserUninstallKey('RustRover 233.8264.22'),
> 
>     ],
> 
>     executableShimPaths: executableShimPathsForJetBrainsIDE('rustrover'),
> 
>     jetBrainsToolboxScriptName: 'rustrover',
> 
>     displayNamePrefixes: ['RustRover '],
> 
>     publishers: ['JetBrains s.r.o.'],
> 
>   },
> 
> ```
> 
> > [!WARNING]  
> 
> > The problem is that this solution is not scalable since the EAP versions are frequently updated.
> 
> 
> 
> @sergiou87, what are your thoughts on the subject?

i think shall we input all of the versions number in it. 👉👈

**Comment by sergiou87** [comment]
Sorry @SnowballXueQiu, we've been quite busy lately 😓 

As @tsvetilian-ty suggested (❤️ ), we would need to add those version numbers. However, until RustRover goes out of beta, we can expect those numbers to change a lot. I discussed this with the team and decided to postpone this until RustRover is officially released and is versioned just like other JetBrains IDEs.

I'm gonna close this PR but, please, feel free to reopen it once that happens!

Sorry again for all this trouble and thank you for your patience 🙇‍♂️

**Comment by SonOfMosiah** [comment]
@sergiou87 RustRover moved out of public preview in May. Would love to see it supported now.

https://blog.jetbrains.com/rust/2024/05/21/rustrover-is-released-and-includes-a-free-non-commercial-option/
